### PR TITLE
TFLM compression changes

### PR DESCRIPTION
**Pull request by ddavis-2015**
@tensorflow/micro

initial header file changes/additions.

bug=#2646

### CODE REVIEW

File: tensorflow/lite/micro/compression.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+static constexpr const char* kCompressionMetadataString = "TFLM_COMPRESSION";
+
+enum class CompressionScheme : uint8_t {
+  kBinQuant,
+};
+
+// TODO(ddavis-2015): pack struct
+struct BinQuantData {
+  static constexpr size_t kMaxBitWidth = 7;
```
Comment (rascani) [inline tensorflow/lite/micro/compression.h:37]: Is there any reason we actually have to constrain this to 7 bits? Theoretically, we could do 8 bit LUTs for f32s?

File: tensorflow/lite/micro/compression.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+//
+
+static constexpr const char* kCompressionMetadataString = "TFLM_COMPRESSION";
+
+enum class CompressionScheme : uint8_t {
+  kBinQuant,
+};
+
+// TODO(ddavis-2015): pack struct
+struct BinQuantData {
```
Comment (rascani) [inline tensorflow/lite/micro/compression.h:36]: Lets just call it LookupTableData. The binning is just a method for determining the lookup table.

File: tensorflow/lite/micro/compression.h (hunk 1 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/compression.h:37]: The index bit width is unrelated to the value table element size.  The associated literature does not use more than 20 bins (5 bit indices).  I'm not currently seeing a use case for a bit width greater than 5.

File: tensorflow/lite/micro/compression.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  const void* value_table;             // Pointer into FlatBuffer Values.
+  uint8_t value_table_channel_stride;  // elements per channel
+  uint8_t compressed_bit_width : 3;    // 1 to 7 bits
+  bool is_per_channel_quantized : 1;   // tensor is per-channel quantized
+  bool use_alternate_axis : 1;         // shape default channel:
+                                       // 0 = first, 1 = last
+  uint8_t reserved : 3;
+};
+
+union CompressionData {
```
[the remark this answers is not in the document]
Comment (ddavis-2015) [reply tensorflow/lite/micro/compression.h:49]: I will change this to a union of pointers.  More compatible with IDEs and debuggers without any explicit type-casting.

File: tensorflow/lite/micro/compression.h (hunk 2 above)
Comment (ddavis-2015) [reply tensorflow/lite/micro/compression.h]: done

**Review by rascani — APPROVED** [review]

**Pull request merged**